src/modules/DocSystem.jl

```diff
@@ -0,0 +1,262 @@
+    m = module_name(m) === v ? module_parent(m) : m
+    Docs.Binding(m, v)
+end
+
+
+## Docstring containers. ##
+
+#
+# `MultiDoc` objects contain a collection of related `DocStr` objects.
+#
+# Here "related" means that they share the same `Binding` and defining module.
+#
+if isdefined(Base.Docs, :MultiDoc)
+    import Base.Docs: MultiDoc
+else
+    immutable MultiDoc
+        order :: Vector{Type}
+        docs  :: ObjectIdDict
+    end
+    MultiDoc() = MultiDoc([], ObjectIdDict())
+end
+
+"""
+Construct a `MultiDoc` object from the provided argument.
+
+Valid inputs are:
+
+- `Markdown.MD`
+- `Docs.FuncDoc`
+- `Docs.TypeDoc`
+
+"""
+function multidoc end
+
+function multidoc(markdown::Markdown.MD)
+    md = MultiDoc()
+    sig = Union{}
+    push!(md.order, sig)
+    md.docs[sig] = docstr(markdown)
+    md
+end
+
+if isdefined(Base.Docs, :FuncDoc)
+    function multidoc(funcdoc::Docs.FuncDoc)
+        md = MultiDoc()
+        append!(md.order, funcdoc.order)
+        for (k, v) in funcdoc.meta
+            md.docs[k] = docstr(v)
+        end
+        md
+    end
+end
+
+if isdefined(Base.Docs, :TypeDoc)
+    function multidoc(typedoc::Docs.TypeDoc)
+        md = MultiDoc()
+        append!(md.order, typedoc.order)
+        if typedoc.main !== nothing
+            unshift!(md.order, Union{})
+            md.docs[Union{}] = docstr(typedoc.main, typedoc.fields)
+        end
+        for (k, v) in typedoc.meta
+            md.docs[k] = docstr(v)
+        end
+        md
+    end
+end
+
+#
+# `DocStr` objects store a raw documentation string and it's parsed form,
+# which is typically a `Markdown.MD` object.
+#
+# Additionally arbitrary metadata in the form of `Symbol => Any` pairs may be
+# stored in the `.data` `Dict`.
+#
+if isdefined(Base.Docs, :DocStr)
+    import Base.Docs: DocStr
+else
+    type DocStr
+        text   :: SimpleVector
+        object :: Nullable
+        data   :: Dict{Symbol, Any}
+    end
+end
+
+"""
+Construct a `DocStr` object from a `Markdown.MD` object.
+
+An optional `Dict` argument provides the docstrings for type fields if any are
+available.
+"""
+function docstr end
+
+function docstr(md::Markdown.MD, fields::Dict)
+    doc = docstr(md)
+    doc.data[:fields] = fields
+    doc
+end
+docstr(md::Markdown.MD) = DocStr(Core.svec(), Nullable(md), Dict())
+
+
+## Converting docstring caches.
+
+"""
+Converts a `0.4`-style docstring cache into a `0.5` one.
+
+The original docstring cache is not modified.
+"""
+function convertmeta(meta::ObjectIdDict)
+    if !haskey(CACHED, meta)
+        docs = ObjectIdDict()
+        for (k, v) in meta
+            if validdoc(k)
+                docs[binding(k)] = multidoc(v)
+            end
+        end
+        CACHED[meta] = docs
+    end
+    CACHED[meta]::ObjectIdDict
+end
+const CACHED = ObjectIdDict()
+
+#
+# Is the documented object `x` valid in the docsystem in `0.5`?
+#
+validdoc(x) = !isa(x, Union{Number, AbstractString, ObjectIdDict})
```

Comment:
I'll probably just inline the `validdoc` definition into the method above, since it's not useful elsewhere. So won't need a docstring after that. Re: the other comment blocks, probably the second `getdocs` can be a docstring I think.

(If you've got any requests for functionality that would be useful to have in this module do let me know.)